libzmix/src/signatures/bbs/signature.rs

```diff
@@ -0,0 +1,281 @@
+use super::keys::{PublicKey, SecretKey};
+use crate::errors::prelude::*;
+use amcl_wrapper::{
+    constants::{GroupG1_SIZE, MODBYTES},
+    extension_field_gt::GT,
+    field_elem::FieldElement,
+    group_elem::{GroupElement, GroupElementVector},
+    group_elem_g1::G1,
+    group_elem_g2::G2,
+};
+
+use amcl_wrapper::field_elem::FieldElementVector;
+use amcl_wrapper::group_elem_g1::G1Vector;
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Signature {
+    pub a: G1,
+    pub e: FieldElement,
+    pub s: FieldElement,
+}
+
+// https://eprint.iacr.org/2016/663.pdf Section 4.3
+impl Signature {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut out = Vec::with_capacity(GroupG1_SIZE + MODBYTES * 2);
+        out.extend_from_slice(self.a.to_bytes().as_slice());
+        out.extend_from_slice(self.e.to_bytes().as_slice());
+        out.extend_from_slice(self.s.to_bytes().as_slice());
+        out
+    }
+
+    pub fn from_bytes(data: &[u8]) -> Result<Signature, BBSError> {
+        let expected = GroupG1_SIZE + MODBYTES * 2;
+        if data.len() != expected {
+            return Err(BBSError::from_kind(BBSErrorKind::SignatureIncorrectSize(
+                data.len(),
+            )));
+        }
+        let mut index = 0;
+        let a = G1::from_bytes(&data[0..GroupG1_SIZE])
+            .map_err(|_| BBSError::from_kind(BBSErrorKind::SignatureValueIncorrectSize))?;
+        index += GroupG1_SIZE;
+        let e = FieldElement::from_bytes(&data[index..(index + MODBYTES)])
+            .map_err(|_| BBSError::from_kind(BBSErrorKind::SignatureValueIncorrectSize))?;
+        index += MODBYTES;
+        let s = FieldElement::from_bytes(&data[index..(index + MODBYTES)])
+            .map_err(|_| BBSError::from_kind(BBSErrorKind::SignatureValueIncorrectSize))?;
+        Ok(Signature { a, e, s })
+    }
+
+    // No committed messages, All messages known to signer.
+    pub fn new(
+        messages: &[FieldElement],
+        signkey: &SecretKey,
+        verkey: &PublicKey,
+    ) -> Result<Self, BBSError> {
+        check_verkey_message(messages, verkey)?;
+        let e = FieldElement::random();
+        let s = FieldElement::random();
+        let b = compute_b_const_time(&G1::new(), verkey, messages, &s, 0);
+        let mut exp = signkey.clone();
+        exp += &e;
+        exp.inverse_mut();
+        let a = b * exp;
```

Comment:
Very unlikely that `a` would be the identity element but maybe we should still check that it's not since the verifies does not accept proof where `a` is identity.